File: Ros2_Practice/Writing_ServiceAndClient/Raw_sercli.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class SrvCliNode : public rclcpp::Node
{
public:
  SrvCliNode() : Node("srv_cli_node")
  {
    server_ = this->create_service<example_interfaces::srv::AddTwoInts>(
      "my_service", std::bind(&SrvCliNode::handle_service, this, std::placeholders::_1, std::placeholders::_2));

    client_ = this->create_client<example_interfaces::srv::AddTwoInts>("my_service");
  }

  void send_request(int a, int b)
  {
    if (!client_->wait_for_service(std::chrono::seconds(1))) {
      RCLCPP_WARN(this->get_logger(), "Waiting for service to appear...");
      return;
    }

    auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    request->a = a;
    request->b = b;

    client_->async_send_request(request, std::bind(&SrvCliNode::response_callback, this, std::placeholders::_1));
  }

private:
  void handle_service(const std::shared_ptr<example_interfaces::srv::AddTwoInts::Request> request,
                      std::shared_ptr<example_interfaces::srv::AddTwoInts::Response> response)
  {
    response->sum = request->a + request->b;
    RCLCPP_INFO(this->get_logger(), "Server responded with sum: %ld", response->sum);
  }

  void response_callback(rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future)
  {
    auto response = future.get();
    RCLCPP_INFO(this->get_logger(), "Client received result: %ld", response->sum);
  }

  rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr server_;
  rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SrvCliNode>());
  rclcpp::shutdown();
  return 0;
}

File: Ros2_Practice/Writing_ServiceAndClient/Raw_sercli.py
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts

class SrvCliNode(Node):
    def __init__(self):
        super().__init__('srv_cli_node')
        
        self.srv = self.create_service(AddTwoInts, 'my_service', self.service_callback)
        self.cli = self.create_client(AddTwoInts, 'my_service')

    def send_request(self, a, b):
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Service not available, waiting...')
        
        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        
        future = self.cli.call_async(request)
        future.add_done_callback(self.response_callback)

    def service_callback(self, request, response):
        response.sum = request.a + request.b
        self.get_logger().info(f'Server logic executed. Sum: {response.sum}')
        return response

    def response_callback(self, future):
        try:
            response = future.result()
            self.get_logger().info(f'Client received response: {response.sum}')
        except Exception as e:
            self.get_logger().error(f'Service call failed: {e}')

def main(args=None):
    rclpy.init(args=args)
    node = SrvCliNode()
    
    # node.send_request(3, 5) # Uncomment to trigger a request
    
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()